pkg/security/api/types.go

```diff
@@ -0,0 +1,88 @@
+package api
+
+import (
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+// PodSpecSubjectReview checks whether a particular user/SA tuple can create the PodSpec.
+type PodSpecSubjectReview struct {
+	unversioned.TypeMeta
+
+	// Spec defines specification for the PodSpecSubjectReview.
+	Spec PodSpecSubjectReviewSpec
+
+	// Status represents the current information/status for the PodSpecSubjectReview.
+	Status PodSpecSubjectReviewStatus
+}
+
+type PodSpecSubjectReviewSpec struct {
+	// PodSpec is the PodSpec to check.
+	PodSpec kapi.PodSpec
+
+	// User is the user you're testing for.
+	// If you specify "User" but not "Group", then is it interpreted as "What if User were not a member of any groups.
+	// If User and Groups are empty, then the check is performed using *only* the ServiceAccountName in the PodSpec.
+	User string
+
+	// Groups is the groups you're testing for.
+	Groups []string
+}
+
+// PodSpecSubjectReviewStatus contains information/status for PodSpecSubjectReview.
+type PodSpecSubjectReviewStatus struct {
+	// AllowedBy is a reference to the rule that allows the PodSpec.  A `nil`, indicates that it was denied.
+	AllowedBy *kapi.ObjectReference
+
+	// Reason is optional.  It indicates why a request was allowed or denied.
+	Reason string
+
+	// PodSpec is the PodSpec after the defaulting is applied.
+	PodSpec kapi.PodSpec
+}
+
+// PodSpecSelfSubjectReview checks whether this user/SA tuple can create the PodSpec.
+type PodSpecSelfSubjectReview struct {
+	unversioned.TypeMeta
+
+	// Spec defines specification the PodSpecSelfSubjectReview.
+	Spec PodSpecSelfSubjectReviewSpec
+
+	// Status represents the current information/status for the PodSpecSelfSubjectReview.
+	Status PodSpecSubjectReviewStatus
+}
+
+// PodSpecSelfSubjectReviewSpec contains specification for PodSpecSelfSubjectReview.
+type PodSpecSelfSubjectReviewSpec struct {
+	// PodSpec is the PodSpec to check.
+	PodSpec kapi.PodSpec
+
+	// ExcludeSelf indicates that the check is performed using *only* the ServiceAccountName in the PodSpec.
+	ExcludeSelf bool
+}
+
+// PodSpecReview checks which service accounts (not users, since that would be cluster-wide) can create the `PodSpec` in question.
+type PodSpecReview struct {
+	unversioned.TypeMeta
+
+	// Spec is the PodSpec to check.  The ServiceAccountName field is ignored for this check.
+	Spec kapi.PodSpec
```

Comment:
I think I want a different spec where I can specify an optional set of service accounts to run the check with.  That would let me eliminate `ExcludeSelf` from the `Self` option and we'd encourage its use strictly for one off pods or we eliminate the API entirely.  I only want to keep it for parity with `PodSpecSubjectReview` which only a project admin should have.